src/Components/Components/src/Forms/InputComponents/InputBase.cs

```diff
@@ -0,0 +1,189 @@
+                    // Assume if it's a nullable type, null/empty inputs should correspond to default(T)
+                    // Then all subclasses get nullable support almost automatically (they just have to
+                    // not reject Nullable<T> based on the type itself).
+                    parsingFailed = false;
+                    CurrentValue = default;
+                }
+                else if (TryParseValueFromString(value, out var parsedValue, out var validationErrorMessage))
+                {
+                    parsingFailed = false;
+                    CurrentValue = parsedValue;
+                }
+                else
+                {
+                    parsingFailed = true;
+
+                    if (_parsingValidationMessages == null)
+                    {
+                        _parsingValidationMessages = new ValidationMessageStore(EditContext);
+                    }
+
+                    _parsingValidationMessages.Add(FieldIdentifier, validationErrorMessage);
+
+                    // Since we're not writing to CurrentValue, we'll need to notify about modification from here
+                    EditContext.NotifyFieldChanged(FieldIdentifier);
+                }
+
+                // We can skip the validation notification if we were previously valid and still are
+                if (parsingFailed || _previousParsingAttemptFailed)
+                {
+                    EditContext.NotifyValidationStateChanged();
+                    _previousParsingAttemptFailed = parsingFailed;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Formats the value as a string. Derived classes can override this to determine the formating used for <see cref="CurrentValueAsString"/>.
+        /// </summary>
+        /// <param name="value">The value to format.</param>
+        /// <returns>A string representation of the value.</returns>
+        protected virtual string FormatValueAsString(T value)
+            => value?.ToString();
+
+        /// <summary>
+        /// Parses a string to create an instance of <typeparamref name="T"/>. Derived classes can override this to change how
+        /// <see cref="CurrentValueAsString"/> interprets incoming values.
+        /// </summary>
+        /// <param name="value">The string value to be parsed.</param>
+        /// <param name="result">An instance of <typeparamref name="T"/>.</param>
+        /// <param name="validationErrorMessage">If the value could not be parsed, provides a validation error message.</param>
+        /// <returns>True if the value could be parsed; otherwise false.</returns>
+        protected virtual bool TryParseValueFromString(string value, out T result, out string validationErrorMessage)
+            => throw new NotImplementedException($"Components that inherit from {nameof(InputBase<T>)} must override " +
+                $"{nameof(TryParseValueFromString)} in order to use {nameof(CurrentValueAsString)}.");
```

Comment:
I'll take a look a the other places where this gets overridden, but a possible enhancement here would be to use `IConvertable` or `TypeConverter` to handle this.

I expect using `TypeConverter` might require more linker stuff.

Reply:
Also, is it really reasonable to do things this way? Is checkbox the only case where we don't need to parse?

Reply:
OK, I agree it would be more conventional to make `TryParseValueFromString` an `abstract` method. Done.